### PR TITLE
systemd-mount: allow to specify an arbitrary string for arg_mount_what when vfs is used

### LINKED ISSUE #6591: systemd-mount wants a valid WHAT parameter for virtual filesystems
### Submission type

 - Bug report

### systemd version the issue has been seen with

234

### Used distribution

custom : buildroot

### In case of bug report: Expected behaviour you didn't see
( /mountpoint exists)

`systemd-mount -t tmpfs none /mountpoint`

should succeed

### In case of bug report: Unexpected behaviour you saw

`systemd-mount -t tmpfs none /mountpoint`

failed and reported the following error message

`Failed to canonicalize path : no such file or directory`

note that entering a similar line in /etc/fstab works correctly
note that using any existing path instead of _none_ works correctly

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #6591.

Cc: @boucman

### CODE REVIEW

**Comment by boucman** [comment]
thx a lot

I had started looking at a fix on the systemd-mount side, but it caused more problems than it fixed. 

I'll test that later

File: src/mount/mount-tool.c [hunk 1]
```diff
@@ -330,7 +330,17 @@ static int parse_argv(int argc, char *argv[]) {
...
                 }
 
-                if (arg_transport == BUS_TRANSPORT_LOCAL) {
+                if (arg_mount_type
+                    && STR_IN_SET(arg_mount_type,
+                                  // virtual file system
+                                  "autofs", "cgroup", "cgroup2", "configfs", "debugfs", "devpts", "devtmpfs", "efivarfs",
+                                  "hugetlbfs", "mqueue", "proc", "pstore", "ramfs", "securityfs", "sysfs", "tmpfs",
+                                  // network file system
+                                  "cifs", "nfs", "nfs4", "smbfs")) {
```
Comment (boucman) [inline src/mount/mount-tool.c]: maybe externalize this set somewhere, so it's define at a single place and easy to update ?

or is that too forward-looking ?

Comment (poettering) [reply src/mount/mount-tool.c]: Uh, I am not sure I am overly enthusiastic about maintaining these lists…

 That said, we already have a function for the network tests in fstype_is_network(). And I figure adding fstype_is_api_vfs() would be acceptable...

Comment (yuwata) [reply src/mount/mount-tool.c]: OK. I will update.

**Comment by boucman** [comment]
Ok I can confirm this fixes the original bug...

**Comment by yuwata** [comment]
Updated.
The function `fstype_is_network()` checks both e.g., `nfs` and `fuse.nfs`. I do not know whether we should also check both `tmpfs` and `fuse.tmpfs`. So, in the new function `fstype_is_api_vfs()`, only fstype like `tmpfs` is checked.

**Comment by poettering** [comment]
> The function fstype_is_network() checks both e.g., nfs and fuse.nfs. I do not know whether we should also check both tmpfs and fuse.tmpfs

For some network file systems there are userspace implementations around, and I figure it doesn't hurt detecting those in blanket way, they way we currently do it. But I don't see anything similar applying to API VFS I must say

**Comment by poettering** [comment]
lgtm

**Comment by yuwata** [comment]
@poettering Thank you for your comment.

**Pull request merged**